### PR TITLE
SDK: allow SDK mismatches if major version matches

### PR DESCRIPTION
**Pull request by mistydemeo**
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/HEAD/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [ ] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/HEAD/Library/Homebrew/test/PATH_spec.rb).
- [x] Have you successfully run `brew style` with your changes locally?
- [x] Have you successfully run `brew tests` with your changes locally?
- [x] Have you successfully run `brew man` locally and committed any changes?

-----

This PR complements #9327 and should be shipped alongside or after it.

@fxcoudert noticed an issue with our current SDK lookup code, described in #9324. @MikeMcQuaid wrote a fix in #9327 which resolves that bug, but leaves a new one: if the OS version and the SDK version are slightly different, Homebrew won't fetch the SDK. In the past that was unlikely, but thanks to the new naming scheme from macOS 11, it's actually quite a bit more common now.

Starting with macOS 11, Apple appears to be planning to increment the major number with every release instead of leaving us with an eternal 10 (or 11). That means that the series number, which in the past was the OS number, now indicates the version of each OS - macOS 11.1, 11.2, etc. are all updates to macOS 11. The Xcode 12.3 beta introduces a `MacOSX11.1` SDK, alongside the minor macOS 11.1 release. It looks like there's a very real chance that the SDK version and OS version will slightly diverge over time, or at least that the SDK version will lag. It's likely that users may have macOS 11.1 installed while using a `MacOSX11.0` SDK, or vice versa. As a result, we need to be able to pick SDKs with a version mismatch, *as long as the major version matches* (and as long as it's macOS 11 or greater).

This PR implements that, and I've tested it using macOS 11.1 running Xcode 12.2 (which ships an `11.0` SDK). Without this PR and with the commit from #9327, we fail to locate the 11.0 SDK:

```
$ sw_vers -productVersion
11.1
$ brew ruby -e 'puts MacOS.sdk_path'

```

With this PR and with the commit from #9327, however, we do find the correct SDK:

```
$ sw_vers -productVersion
11.1
$ brew ruby -e 'puts MacOS.sdk_path'
/Applications/Xcode-beta.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX11.0.sdk
```

With the two of these together we should cover the weird edge cases in SDK handling. As noted by @fxcoudert, however, we have a few other edge cases outside SDK selection to patch up as well.

cc @MikeMcQuaid

### CODE REVIEW

**Review by fxcoudert — APPROVED** [review]

File: Library/Homebrew/os/mac/sdk.rb
Comment (MikeMcQuaid) [inline Library/Homebrew/os/mac/sdk.rb:62]: Can this use the :big_sur symbol rather than using 11? This should also handle the case where sdk is nil.

File: Library/Homebrew/os/mac/sdk.rb
Comment (MikeMcQuaid) [inline Library/Homebrew/os/mac/sdk.rb:66]: My understanding from this is it already allows mismatched SDK versions. Can you explain what this covers and does not that the new code does?

Relatedly: could this be refactored to not use unless && or || (either use if or two statements).  Not your code I know but I find it really hard to parse this as-is.

**Review by MikeMcQuaid — APPROVED** [review]
Thanks for this! Once comments are addressed feel free to merge. I won’t be around to supervise #9327 going out so feel free to cherry pick that in here and merge them together. Thanks!

File: Library/Homebrew/os/mac/sdk.rb
Comment (mistydemeo) [reply Library/Homebrew/os/mac/sdk.rb:66]: This allows picking an SDK which is *newer* than the one for the active OS, while explicitly disallowing SDKs *older* than the active OS unless explicitly requested. This handles the case where newer Xcode versions don't ship an SDK for the current OS and instead ship one for the latest OS, which we've seen often. For example, it's not uncommon for users on macOS 10.14 to only have an SDK for macOS 10.15, and so forth.

The case that the new case covers is cases where you have an *older* SDK than your current OS, but the only difference is the minor version and not the OS version. That's a case that can't occur before Big Sur.

File: Library/Homebrew/os/mac/sdk.rb
Comment (mistydemeo) [reply Library/Homebrew/os/mac/sdk.rb]: Done!

File: Library/Homebrew/os/mac/sdk.rb
Comment (MikeMcQuaid) [reply Library/Homebrew/os/mac/sdk.rb:66]:
```suggestion
        return if v.blank? && sdk.present? && sdk.version < OS::Mac.version
```

**Comment by mistydemeo** [comment]
Thanks for merging this for me. Planned to ship it over the weekend but after a couple pages I ended up not having the time. 😅

**Comment by mstange** [comment]
(Initially commented on the wrong PR)

I just updated to homebrew 2.6.0 (which includes this fix) and I'm still having trouble.
I am on the 11.1 Beta and have Xcode 12.2 which includes the `MacOSX11.0.sdk`. Running `brew doctor` still gives me the following message:

```
Warning: Your CLT does not support macOS 11.
It is either outdated or was modified.
```

Was this PR intended to fix this?

**Comment by MikeMcQuaid** [comment]
@mstange Can I see your `brew config` and fun `brew doctor` output, please?

**Comment by mstange** [comment]
I re-downloaded the Command Line Tools for Xcode 12.2 as instructed by `brew doctor`, and that fixed the issue. Sorry about the troubles.

**Pull request merged**
